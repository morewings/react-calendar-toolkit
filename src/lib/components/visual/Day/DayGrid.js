import React from 'react';
import PropTypes from 'prop-types';
import classes from './Day.module.css';

const DayGrid = ({children}) => <div className={classes.grid}>{children}</div>;

DayGrid.propTypes = {
  children: PropTypes.node.isRequired,
};

export default DayGrid;
