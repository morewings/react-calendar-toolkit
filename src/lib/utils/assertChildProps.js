import React, {Fragment} from 'react';

export const trackProps = jest.fn();

export const DummyComponent = props => {
  trackProps(props);
  return <Fragment>Child</Fragment>;
};
