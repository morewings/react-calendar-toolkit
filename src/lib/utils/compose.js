const compose = (...fns) => (...args) =>
  fns.reduceRight(
    (params, f) => (Array.isArray(params) ? f(...params) : f(params)),
    args
  );

export default compose;
