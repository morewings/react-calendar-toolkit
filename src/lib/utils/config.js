export default Object.freeze({
  supportedPrecisions: Object.freeze([
    'year',
    'month',
    'day',
    'hour',
    'minute',
  ]),
});
