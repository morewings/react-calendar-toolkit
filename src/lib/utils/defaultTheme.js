const baseColor = '#43a047';
const gray = 'rgba(0, 0, 0, 0.38)';
const borderRadius = '4px';

const defaultTheme = {
  /** Set font for datepicker */
  fontPrimary: 'Helvetica Neue',
  /** Font, which is applied, if --fontPrimary is not available */
  fontSecondary: 'Arial',
  /** Fallback font if two previous are unavailable */
  fontFallback: 'sans-serif',
  fontSize: '14px',
  fontWeight: '400',
  /** Applied to the title */
  fontSizeSmall: '10px',
  /** Applied to the header text */
  fontSizeBig: '24px',
  /** Set inner width */
  calendarWidth: '240px',
  /** Calculate total width */
  datePickerWidth: 'calc(var(--calendarWidth) + var(--innerPadding) * 2)',
  selectorHeight: '36px',
  calendarHeight: '200px',
  /** Set padding between elements */
  innerPadding: '12px',
  /** Day, month, year entries height */
  entryHeight: '32px',
  entryMargin: '1px',
  datepickerBorderRadius: borderRadius,
  entryBorderRadius: borderRadius,
  /** Background color of Calendar */
  bgColor: 'white',
  /** Base color for text */
  textColor: 'black',
  headerBgColor: baseColor,
  weekendTextColor: 'red',
  headerTextColor: 'white',
  borderColor: gray,
  modalBgColor: gray,
  weekDayColor: gray,
  entryHoverColor: 'rgba(0, 0, 0, 0.08)',
  entrySelectedColor: baseColor,
  entryHighlightColor: 'lightblue',
  inputWidth: '90px',
  inputPadding: '8px',
  inputBorderRadius: borderRadius,
  inputBorderColor: gray,
  inputHighlightColor: 'lightblue',
};

export default defaultTheme;
