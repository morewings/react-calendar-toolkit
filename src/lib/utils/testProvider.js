import React from 'react';
import PropTypes from 'prop-types';
import {Provider as LocaleProvider} from 'lib/features/locale/withLocaleContext';
import {Provider as ThemeProvider} from 'lib/features/theme/withTheme';
import {Provider as ModalProvider} from 'lib/features/modal/withModalProvider';
import {Provider as DatepickerProvider} from 'lib/features/datepicker/withDatepickerProvider';

export const DatepickerMockProvider = props => (
  <LocaleProvider {...props}>
    <ThemeProvider>
      <DatepickerProvider>{props.children}</DatepickerProvider>
    </ThemeProvider>
  </LocaleProvider>
);

export const InputMockProvider = props => (
  <LocaleProvider {...props}>
    <ThemeProvider>
      <ModalProvider>
        <DatepickerProvider>{props.children}</DatepickerProvider>
      </ModalProvider>
    </ThemeProvider>
  </LocaleProvider>
);

DatepickerMockProvider.propTypes = {
  children: PropTypes.node.isRequired,
};

InputMockProvider.propTypes = {
  children: PropTypes.node.isRequired,
};
