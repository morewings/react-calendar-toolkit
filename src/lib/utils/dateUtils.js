import {
  formatWithOptions,
  isSameMonth,
  isSameYear,
  toDate,
  getTime,
  addMonths,
  addDays,
  addYears,
  isWithinInterval,
  isWeekend,
  isSameDay,
  isSameHour,
  getDate,
  startOfMonth,
  endOfMonth,
} from 'date-fns/fp';
import compose from 'lib/utils/compose';

/**
 * Unix timestamp
 * @typedef {number} UnixTimestamp
 */

/**
 * Date union type may be Date or Unix timestamp
 * @typedef {(Date|UnixTimestamp)} DateUnion
 */

/**
 * @function
 * @name formatDateWithLocale
 * @description Returns formatted date
 * @param {Object} options - Options object with locale
 * @param {string} pattern - Formatting pattern
 * @param {Date} date - Date object to apply format
 * @return {string} Formatted date
 */
export const formatDateWithLocale = formatWithOptions;

/**
 * @function
 * @name checkIsSameMonth
 * @description Checks if two dates belong to the same month
 * @param {DateUnion} dateLeft - Date to check
 * @param {DateUnion} dateRight - Date to check
 * @return {boolean}
 */
export const checkIsSameMonth = isSameMonth;

/**
 * @function
 * @name checkIsSameYear
 * @description Checks if two dates belong to the same year
 * @param {DateUnion} dateLeft - Date to check
 * @param {DateUnion} dateRight - Date to check
 * @return {boolean}
 */
export const checkIsSameYear = isSameYear;

/**
 * @function
 * @name checkIsSameDay
 * @description Checks if two dates belong to the same day
 * @param {DateUnion} dateLeft - Date to check
 * @param {DateUnion} dateRight - Date to check
 * @return {boolean}
 */
export const checkIsSameDay = isSameDay;

/**
 * @function
 * @name checkIsSameHour
 * @description Checks if two dates belong to the same hour
 * @param {DateUnion} dateLeft - Date to check
 * @param {DateUnion} dateRight - Date to check
 * @return {boolean}
 */
export const checkIsSameHour = isSameHour;

/**
 * @function
 * @name convertToTimestamp
 * @description Converts provided date to Unix timestamp
 * @param {Date} date - Date to convert
 * @return {UnixTimestamp} Unix timestamp
 */
export const convertToTimestamp = date => getTime(date);

/**
 * @function
 * @name convertToDate
 * @description Converts provided Unix timestamp to Date object
 * @param {number} timestamp - Date to convert
 * @return {Date}
 */
export const convertToDate = timestamp => toDate(timestamp);

/**
 * @function
 * @name convertToDayOfMonth
 * @description Converts provided Unix timestamp or Date to day of month number
 * @param {DateUnion} timestamp - Timestamp or Date to convert
 * @return {number} Day of the month
 */
export const convertToDayOfMonth = timestamp => getDate(timestamp);

/**
 * @function
 * @name checkIsWithinInterval
 * @description Checks if date belongs to the interval
 * @typedef {Object} Interval
 * @property {Date} start - Start of interval
 * @property {Date} end - End of interval
 * @param {DateUnion} date - Date to check
 * @param {Interval} interval - Time range to check
 * @return {boolean}
 */
export const checkIsWithinInterval = isWithinInterval;

/**
 * @function
 * @name checkIsWeekend
 * @description Checks if date belongs to the weekend
 * @param {DateUnion} date - Date to check
 * @return {boolean}
 */
export const checkIsWeekend = date => isWeekend(date);

/**
 * @function
 * @name matchDatesWithPrecision
 * @description Checks if dates are the same with given precision, eg belong to same day|month|year
 * @param {string} precision - Date to check
 * @param {DateUnion} dateLeft - Date to check
 * @param {DateUnion} dateRight - Date to check
 * @return {boolean}
 */
export const matchDatesWithPrecision = (precision, dateLeft, dateRight) =>
  ({
    day: checkIsSameDay,
    month: checkIsSameMonth,
    year: checkIsSameYear,
  }[precision](dateLeft, dateRight));

/**
 * @function
 * @name ceilMonth
 * @description Gets end of month date
 * @param {DateUnion} date
 * @return {Date}
 */
export const ceilMonth = endOfMonth;

/**
 * @function
 * @name floorMonth
 * @description Gets start of month date
 * @param {DateUnion} date
 * @return {Date}
 */
export const floorMonth = startOfMonth;

/**
 * @function
 * @name addMonth
 * @description Adds months
 * @param {DateUnion} date - Date or Unix timestamp
 * @param {number} amount - Amount of months to add
 * @return {UnixTimestamp}
 */
export const addMonth = (date, amount) =>
  compose(convertToTimestamp, addMonths)(amount, date);

/**
 * @function
 * @name subMonth
 * @description Subtracts months
 * @param {DateUnion} date - Date or Unix timestamp
 * @param {number} amount - Amount of months to subtract
 * @return {UnixTimestamp}
 */
export const subMonth = (date, amount) =>
  compose(convertToTimestamp, addMonths)(-amount, date);

/**
 * @function
 * @name addDay
 * @description Adds days
 * @param {DateUnion} date - Date or Unix timestamp
 * @param {number} amount - Amount of days to add
 * @return {UnixTimestamp}
 */
export const addDay = (date, amount) =>
  compose(convertToTimestamp, addDays)(amount, date);

/**
 * @function
 * @name subDay
 * @description Subtracts days
 * @param {DateUnion} date - Date or Unix timestamp
 * @param {number} amount - Amount of days to subtract
 * @return {UnixTimestamp}
 */
export const subDay = (date, amount) =>
  compose(convertToTimestamp, addDays)(-amount, date);

/**
 * @function
 * @name addYear
 * @description Adds years
 * @param {DateUnion} date - Date or Unix timestamp
 * @param {number} amount - Amount of years to add
 * @return {UnixTimestamp}
 */
export const addYear = (date, amount) =>
  compose(convertToTimestamp, addYears)(amount, date);

/**
 * @function
 * @name subYear
 * @description Subtracts years
 * @param {DateUnion} date - Date or Unix timestamp
 * @param {number} amount - Amount of days to subtract
 * @return {UnixTimestamp}
 */
export const subYear = (date, amount) =>
  compose(convertToTimestamp, addYears)(-amount, date);
