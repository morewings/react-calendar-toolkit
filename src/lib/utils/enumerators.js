import {
  startOfYear,
  getYear,
  getHours as getHourValue,
  getMinutes as getMinuteValue,
  addYears,
  addDays,
  toDate,
  differenceInDays,
  endOfWeekWithOptions,
  startOfWeekWithOptions,
  getDate,
  startOfDay,
  endOfDay,
  min,
  max,
  eachHourOfInterval,
  startOfHour,
  endOfHour,
  isBefore,
  addMinutes,
} from 'date-fns/fp';
import {addMonth, ceilMonth, floorMonth} from 'lib/utils/dateUtils';
import curry from 'lib/utils/curry';
import getDaytimeLabels from 'lib/utils/getDaytimeLabels';

/**
 * Item name
 * @typedef {Object} ItemName
 * @property {string} [narrow] - Narrow name of item - 'S'
 * @property {string} [short] - Short name of item - 'Su'
 * @property {string} [abbreviated] - Abbreviated name of item - 'Sun'
 * @property {string} [wide] - Full name of item - 'Sunday'
 * @property {number} [numeric] - Numeric name of item - 1
 */

/**
 * Item description
 * @typedef {Object} ItemDescription
 * @property {ItemName} name - Item name vocabulary
 * @property {Date} date - starting date of the item
 */

/**
 * Date union type may be Date or Unix timestamp
 * @typedef {(Date|number)} DateUnion
 */

/**
 * @function
 * @name getWeekDayNames
 * @description Returns array of weekday names
 * @param {Object} locale - Date-fns locale object
 * @return {Array.<ItemName>}
 */
export const getWeekDayNames = locale =>
  [...Array(7).keys()].map(i => {
    const dayNumber =
      i + locale.options.weekStartsOn === 7
        ? 0
        : i + locale.options.weekStartsOn;
    return {
      short: locale.localize.day(dayNumber, {width: 'short'}),
      narrow: locale.localize.day(dayNumber, {width: 'narrow'}),
      abbreviated: locale.localize.day(dayNumber, {width: 'abbreviated'}),
      wide: locale.localize.day(dayNumber, {width: 'wide'}),
      numeric: i,
    };
  });

/**
 * @function
 * @name getYears
 * @description Returns collection year description objects in the provided range
 * @param {DateUnion} startDate - Start date
 * @param {DateUnion} endDate - End date
 * @return {Array.<ItemDescription>}
 */
const getYears = (startDate, endDate) => {
  const startYear = getYear(startDate);
  const endYear = getYear(endDate) + 1;
  const years = [...Array(endYear - startYear).keys()];
  return years.map((year, i) => ({
    name: {
      numeric: startYear + i,
    },
    date: addYears(i, startDate),
  }));
};

const getYearsCurried = curry(getYears);

export {getYearsCurried as getYears};

/**
 * @function
 * @name getMonths
 * @description Returns collection of month description objects based on provided date
 * @param {Object} locale - Date-fns locale object
 * @param {DateUnion} date - Date or Unix timestamp
 * @return {Array.<ItemDescription>}
 */
const getMonths = (locale, date) => {
  const year = startOfYear(date);
  const monthNames = [...Array(12).keys()].map(i => ({
    wide: locale.localize.month(i, {width: 'wide'}),
    abbreviated: locale.localize.month(i, {width: 'abbreviated'}),
    narrow: locale.localize.month(i, {width: 'narrow'}),
    numeric: i,
  }));
  return monthNames.map((name, i) => ({
    name,
    date: toDate(addMonth(year, i)),
  }));
};

const getMonthsCurried = curry(getMonths);

export {getMonthsCurried as getMonths};

/**
 * @function
 * @name getDays
 * @description Returns collection of day description objects based on provided date
 * @param {Object} locale - Date-fns locale object
 * @param {DateUnion} timestamp - Unix timestamp or Date object
 * @return {Array.<ItemDescription>}
 */
const getDays = (locale, timestamp) => {
  const monthStart = floorMonth(timestamp);
  const monthEnd = ceilMonth(monthStart);
  const startDate = startOfWeekWithOptions({locale}, monthStart);
  const endDate = endOfWeekWithOptions({locale}, monthEnd);
  const cellLength = differenceInDays(startDate, endDate);
  return new Array(cellLength + 1).fill('').map((_, i) => ({
    date: addDays(i, startDate),
    name: {
      numeric: getDate(addDays(i, startDate)),
    },
  }));
};

const getDaysCurried = curry(getDays);

export {getDaysCurried as getDays};

/**
 * @function
 * @name getHours
 * @description Returns collection of hours Date objects based on provided date and interval
 * @param {Object} locale - Date-fns locale object
 * @param {DateUnion} timestamp - Unix timestamp or Date object
 * @param {DateUnion} startDate - Unix timestamp or Date object
 * @param {DateUnion} endDate - Unix timestamp or Date object
 * @return {Array.<ItemDescription>}
 */
export const getHours = (locale, timestamp, startDate, endDate) => {
  const {timeFormat, amLabel, pmLabel} = getDaytimeLabels(locale);
  const HALF_DAY = 12;
  const hourStart = max([startOfDay(timestamp), startDate]);
  const hourEnd = min([endOfDay(timestamp), endDate]);
  const checkIsAm = date => getHourValue(date) <= HALF_DAY;
  const getLocalizedHourName = date => {
    if (timeFormat === '24') {
      return getHourValue(date);
    }
    if (timeFormat === '12') {
      return checkIsAm(date)
        ? getHourValue(date)
        : getHourValue(date) - HALF_DAY;
    }
    return null;
  };
  return eachHourOfInterval({start: hourStart, end: hourEnd}).map(date => ({
    date,
    name: getLocalizedHourName(date),
    daytimeLabel: checkIsAm(date) ? amLabel : pmLabel,
  }));
};

export const getMinutes = (timestamp, startDate, endDate) => {
  const minuteStart = max([startOfHour(timestamp), startDate]);
  const minuteEnd = min([endOfHour(timestamp), endDate]);
  const result = [];
  let minute = minuteStart;
  while (isBefore(minuteEnd, minute)) {
    // eslint-disable-next-line fp/no-mutating-methods
    result.push(minute);
    // eslint-disable-next-line fp/no-mutation
    minute = addMinutes(1, minute);
  }
  return result.map(date => ({
    date,
    name: getMinuteValue(date),
  }));
};
