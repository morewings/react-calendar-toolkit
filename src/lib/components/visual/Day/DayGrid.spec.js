import React from 'react';
import {DatepickerMockProvider as Provider} from 'lib/utils/testProvider';
import {render} from '@testing-library/react';
import DayGrid from './DayGrid';

const Mock = () => <div>Mock</div>;

describe('DayGrid', () => {
  it('renders', () => {
    const {asFragment} = render(
      <DayGrid>
        <Mock />
      </DayGrid>,
      {
        wrapper: ({children}) => <Provider>{children}</Provider>,
      }
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
