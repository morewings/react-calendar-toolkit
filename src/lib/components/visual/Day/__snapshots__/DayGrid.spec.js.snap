// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DayGrid renders 1`] = `
<DocumentFragment>
  <div
    class="foo"
  >
    <div>
      Mock
    </div>
  </div>
</DocumentFragment>
`;
